nav2_controller/src/nav2_controller.cpp

```diff
@@ -0,0 +1,204 @@
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<std_srvs::srv::Empty::Request>/*req*/,
+  std::shared_ptr<std_srvs::srv::Empty::Response>/*res*/)
+{
+  startup();
+}
+
+void
+Nav2Controller::shutdownCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<std_srvs::srv::Empty::Request>/*req*/,
+  std::shared_ptr<std_srvs::srv::Empty::Response>/*res*/)
+{
+  shutdown();
+}
+
+void
+Nav2Controller::pauseCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<std_srvs::srv::Empty::Request>/*req*/,
+  std::shared_ptr<std_srvs::srv::Empty::Response>/*res*/)
+{
+  pause();
+}
+
+void
+Nav2Controller::resumeCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<std_srvs::srv::Empty::Request>/*req*/,
+  std::shared_ptr<std_srvs::srv::Empty::Response>/*res*/)
+{
+  resume();
+}
+
+void
+Nav2Controller::createLifecycleServiceClients()
+{
+  message("Creating and initializing lifecycle service clients");
+  for (auto & node_name : node_names_) {
+    node_map_[node_name] =
+      std::make_shared<LifecycleServiceClient>(node_name, service_client_node_);
+  }
+}
+
+void
+Nav2Controller::destroyLifecycleServiceClients()
+{
+  message("Destroying lifecycle service clients");
+  for (auto & kv : node_map_) {
+    kv.second.reset();
+  }
+}
+
+void
+Nav2Controller::changeStateForAllNodes(std::uint8_t transition)
+{
+  for (const auto & kv : node_map_) {
+    if (!kv.second->change_state(transition)) {
+      std::cerr << ANSI_COLOR_RED "Failed to change state for " << kv.first.c_str() <<
+        ANSI_COLOR_RESET;
+      return;
+    }
+  }
+}
+
+void
+Nav2Controller::bringupNode(const std::string & node_name)
+{
+  message(std::string("Configuring and activating ") + node_name);
+  node_map_[node_name]->change_state(Transition::TRANSITION_CONFIGURE);
+  node_map_[node_name]->change_state(Transition::TRANSITION_ACTIVATE);
+}
+
+void
+Nav2Controller::shutdownAllNodes()
+{
+  message("Deactivate, cleanup, and shutdown nodes");
+  changeStateForAllNodes(Transition::TRANSITION_DEACTIVATE);
+  changeStateForAllNodes(Transition::TRANSITION_CLEANUP);
+  changeStateForAllNodes(Transition::TRANSITION_UNCONFIGURED_SHUTDOWN);
+}
+
+void
+Nav2Controller::startup()
+{
+  message("Starting the system bringup...");
+  createLifecycleServiceClients();
+  for (auto & node_name : node_names_) {
+    bringupNode(node_name);
+  }
+  message("The system is active");
+}
+
+void
+Nav2Controller::shutdown()
+{
+  message("Shutting down the system...");
+  shutdownAllNodes();
+  destroyLifecycleServiceClients();
+  message("The system has been sucessfully shut down");
+}
+
+void
+Nav2Controller::pause()
+{
+  message("Pausing the system...");
+  for (const auto & kv : node_map_) {
+    if (!kv.second->change_state(Transition::TRANSITION_DEACTIVATE) ||
+      !kv.second->change_state(Transition::TRANSITION_CLEANUP))
```

Comment:
I'm wondering if I shouldn't just delete Pause and Resume for now so that we can discuss and just leave Startup and Shutdown. We can then discuss how this should work with the overall system in mind - parameters, log files, etc.